libs/framework/src/bundle_archive.c

```diff
@@ -28,644 +28,415 @@
-		bundleArchive_destroy(archive);
-	}
-
-	framework_logIfError(fw->logger, status, error, "Could not create archive");
-
-	return status;
+static void celix_bundleArchive_storeBundleStateProperties(bundle_archive_pt archive) {
+    celix_properties_t* bundleStateProperties = celix_properties_create();
+    if (bundleStateProperties != NULL) {
+        celixThreadMutex_lock(&archive->lock);
+        //set/update bundle cache state properties
+        celix_properties_setLong(bundleStateProperties, CELIX_BUNDLE_ARCHIVE_BUNDLE_ID_PROPERTY_NAME, archive->id);
+        celix_properties_set(bundleStateProperties, CELIX_BUNDLE_ARCHIVE_LOCATION_PROPERTY_NAME, archive->location);
+        celix_properties_set(bundleStateProperties, CELIX_BUNDLE_ARCHIVE_SYMBOLIC_NAME_PROPERTY_NAME,
+                             archive->bundleSymbolicName);
+        celix_properties_set(bundleStateProperties, CELIX_BUNDLE_ARCHIVE_VERSION_PROPERTY_NAME, archive->bundleVersion);
+
+        //save bundle cache state properties
+        celix_properties_store(bundleStateProperties, archive->savedBundleStatePropertiesPath,
+                               "Bundle State Properties");
+        celixThreadMutex_unlock(&archive->lock);
+    }
+    celix_properties_destroy(bundleStateProperties);
 }
 
-celix_status_t bundleArchive_create(celix_framework_t* fw, const char *archiveRoot, long id, const char * location, const char *inputFile, bundle_archive_pt *bundle_archive) {
-	celix_status_t status = CELIX_SUCCESS;
-	char *error = NULL;
-	bundle_archive_pt archive = NULL;
-
-	if (*bundle_archive != NULL) {
-		status = CELIX_ILLEGAL_ARGUMENT;
-		error = "bundle_archive_pt must be NULL";
-	} else {
-		archive = (bundle_archive_pt) calloc(1,sizeof(*archive));
-		if (archive == NULL) {
-			status = CELIX_ENOMEM;
-		} else {
-			status = linkedList_create(&archive->revisions);
-			if (status == CELIX_SUCCESS) {
-				archive->fw = fw;
-				archive->id = id;
-				archive->location = strdup(location);
-				archive->archiveRootDir = NULL;
-				archive->archiveRoot = strdup(archiveRoot);
-				archive->refreshCount = -1;
-				time(&archive->lastModified);
-
-				status = bundleArchive_initialize(archive);
-				if (status == CELIX_SUCCESS) {
-					status = bundleArchive_revise(archive, location, inputFile);
-
-					if (status == CELIX_SUCCESS) {
-						*bundle_archive = archive;
-					}
-					else{
-						bundleArchive_closeAndDelete(archive);
-					}
-				}
-			}
-		}
-	}
-
-	if(status != CELIX_SUCCESS && archive != NULL){
-		bundleArchive_destroy(archive);
-	}
+celix_status_t celix_bundleArchive_extractBundle(
+        bundle_archive_t* archive,
+        const char* revisionRoot,
+        const char* bundleUrl,
+        const struct timespec* revisionModificationTime) {
+    celix_status_t status = CELIX_SUCCESS;
+    bool extractBundle = true;
+
+    //check if bundle location is newer than current revision
+    if (celix_utils_fileExists(revisionRoot)) {
```

Comment:
```suggestion
    if (revisionModificationTime->tv_sec != 0 || revisionModificationTime->tv_nsec != 0) {
```
The original check always return true, since `archive->resourceCacheRoot` has already been created.
I suggest the above check instead.